### PR TITLE
Adding sentry to code-intel services

### LINKED ISSUE #27591: Add Sentry to code-intel
Adding Sentry to code-intel services to better track our errors.

### PR DESCRIPTION
**Pull request by Numbers88s**
## 🤔  Description

Adding sentry to code-intel services 

You can now add different hubs to Sentry client: see [here](https://github.com/sourcegraph/sourcegraph/pull/27592/files#diff-85a2e9591a77140b767e8d355d634abf7c4ece2d56c7b400693ebeac87478aed)

## 🔖  Relevant tickets

Fixes #27591 

## 🚀  GIF
_What gif best describes this PR or how it makes you feel?_

shortcut

### CODE REVIEW

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 1]
```diff
@@ -353,6 +354,7 @@ func NewSchema(
 	dotcom DotcomRootResolver,
 	searchContexts SearchContextsResolver,
 	orgRepositoryResolver OrgRepositoryResolver,
+	codeintelSentry *sentry.Hub,
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/graphqlbackend.go:357]: this is unused?

Comment (eseliger) [reply cmd/frontend/graphqlbackend/graphqlbackend.go:357]: lovely :)

Comment (Numbers88s) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Done 👍 , thanks @eseliger

File: internal/sentry/hub.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+	conf.Watch(func() {
+		newDsn := onChange(conf)
+		if newDsn != "" && newDsn != client.Options().Dsn {
+			c, err := initClient(newDsn)
+			if err != nil {
+				log15.Error("sentry.dsn.initClient", "error", err)
+				return
+			}
+			h.Hub = sentry.NewHub(c, sentry.NewScope())
```
Comment (eseliger) [inline internal/sentry/hub.go:54]: Is this concurrency safe?

File: internal/sentry/hub.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	h.captureError(err, sentry.LevelError, tags)
+}
+
+// CapturePanic does same thing as CaptureError, and adds additional tags to
+// mark the report as "fatal" level.
+func (h *Hub) CapturePanic(err error, tags map[string]string) {
+	h.captureError(err, sentry.LevelFatal, tags)
+}
+
+func (h *Hub) Recoverer(handler http.Handler) http.Handler {
```
Comment (eseliger) [inline internal/sentry/hub.go:93]:
```suggestion
// Recoverer wraps handler and catches any panics raised from it. They will be reported to sentry and the request is finished with a 500 response rate.
func (h *Hub) Recoverer(handler http.Handler) http.Handler {
```

File: internal/sentry/hub.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if r := recover(); r != nil {
+				err := errors.Errorf("handler panic: %v", errors.Safe(r))
+				h.CapturePanic(err, nil)
+
+				log15.Error("recovered from panic", "error", err)
+				debug.PrintStack()
+
+				w.WriteHeader(http.StatusInternalServerError)
```
Comment (eseliger) [inline internal/sentry/hub.go:103]: is it safe to call `WriteHeader` multiple times? Otherwise it might be that `handler` already did this.

**Review by eseliger — APPROVED** [review]
The non codeintel-specific stuff LGTM.

File: enterprise/internal/codeintel/stores/dbstore/observability.go [hunk 5]
```diff
@@ -90,6 +90,9 @@ func newOperations(observationContext *observation.Context, metrics *metrics.RED
 			Name:              fmt.Sprintf("codeintel.dbstore.%s", name),
 			MetricLabelValues: []string{name},
 			Metrics:           metrics,
+			ErrorFilter: func(err error) observation.ErrorFilterBehaviour {
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/dbstore/observability.go:93]: Do we want to do the same thing for lsifstore, gitserver client, upload store?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go [hunk 6]
```diff
@@ -77,17 +86,25 @@ func (r *Resolver) LSIFUploadByID(ctx context.Context, id graphql.ID) (gql.LSIFU
...
 	}
 
-	return NewUploadResolver(r.db, r.resolver, upload, prefetcher, r.locationResolver), nil
+	return NewUploadResolver(r.db, r.resolver, upload, prefetcher, r.locationResolver, r.observationContext.lsifUploadByID), nil
 }
 
 // 🚨 SECURITY: dbstore layer handles authz for GetUploads
-func (r *Resolver) LSIFUploads(ctx context.Context, args *gql.LSIFUploadsQueryArgs) (gql.LSIFUploadConnectionResolver, error) {
+func (r *Resolver) LSIFUploads(ctx context.Context, args *gql.LSIFUploadsQueryArgs) (_ gql.LSIFUploadConnectionResolver, err error) {
+	// ctx, endObservation := r.observationContext.lsifUploads.With(ctx, &err, observation.Args{})
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go:94]: Why is this one commented out?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go [hunk 7]
```diff
@@ -403,7 +478,12 @@ func (r *Resolver) DeleteCodeIntelligenceConfigurationPolicy(ctx context.Context
 }
 
 // 🚨 SECURITY: Only entrypoint is within the repository resolver so the user is already authenticated
-func (r *Resolver) IndexConfiguration(ctx context.Context, id graphql.ID) (gql.IndexConfigurationResolver, error) {
+func (r *Resolver) IndexConfiguration(ctx context.Context, id graphql.ID) (_ gql.IndexConfigurationResolver, err error) {
+	_, endObservation := r.observationContext.indexConfiguration.With(ctx, &err, observation.Args{LogFields: []log.Field{
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go:482]: We should be consistent about the ordering of autoIndexingEnabled/observation bounds. I see this flipped a few times in this file.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go [hunk 8]
```diff
@@ -50,7 +60,14 @@ func (r *configurationPolicyResolver) RepositoryPatterns() *[]string {
 	return r.configurationPolicy.RepositoryPatterns
 }
 
-func (r *configurationPolicyResolver) Type() (gql.GitObjectType, error) {
+func (r *configurationPolicyResolver) Type() (_ gql.GitObjectType, err error) {
+	_, endObservation := r.op.With(context.Background(), &err, observation.Args{LogFields: []log.Field{
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go:64]: Is there a reason we do it for Type but not Pattern, etc?

**Comment by efritz** [comment]
Hold on... some stuff is just starting to click here 🤦 .

**Comment by efritz** [comment]
FYI, talked with @Strum355 in DM and he seemed to have an idea he might be tackling in this PR. Not sure though, will look for an update later.

File: internal/sentry/hub.go (hunk 4 above)
Comment (Strum355) [reply internal/sentry/hub.go:103]: It is not safe, we're just mirroring what the existing Recoverer did here, most of this file is just a mirror onto a type

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go (hunk 8 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go:64]: Pattern etc dont return an error, so its not possible for them to cause a sentry error

File: internal/sentry/hub.go (hunk 2 above)
Comment (Strum355) [reply internal/sentry/hub.go:54]: It is not, but the amount of times I expect this to happen, it should be a non-issue.

File: enterprise/internal/codeintel/stores/dbstore/observability.go (hunk 5 above)
Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/observability.go:93]: lsifstore (and Im assuming others) get an observation context that doesnt have hub attached to it https://github.com/sourcegraph/sourcegraph/pull/27592/files#diff-260750d37412ed77a4855e3c83f84695f3ac05a6f722a00f9ed25faac6032b78R52-R56
Do we want to attach hub to it? Why did we _not_ do that yet?

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/observability.go:93]: We just haven't added sentry in other places, I think. Would be cool to see all errors coming from a particular store as well if that's doable.

Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/observability.go:93]: It should catch errors just by virtue of using internal/observation, so it should be :ok: ?

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/observability.go:93]: How does it send errors w/o a hub?

Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/observability.go:93]: Im gonna exclude them for now just to minimize the blast radius

**Pull request merged**